(PUP-8378) Reject MD5 related digest algorithms in FIPS mode

joshcooper opened the pull request:
FIPS enabled agents will segfault if they attempt MD5 related operations. The `digest_algorithm` is used for filebucketing, and is the default checksum type for file resources. The `supported_checksum_types` setting is used when requesting a catalog, so that the server has an opportunity to inline file metadata in the static catalog.

This PR does not attempt to prevent the user from setting `md5` in puppet settings, as that is a more general problem in puppet.

Supersedes PR #6581

puppetcla commented:
CLA signed by all contributors.

kris-bosland commented on lib/puppet/type/file/checksum.rb:
```diff
@@ -15,6 +15,12 @@
   end
 
+  validate do |value|
+    if Puppet::Util::Platform.fips_enabled? && (value == :md5 || value == :md5lite)
```
The test in lib/puppet/defaults.rb#970 is grep(/^md5/), should these tests be the same?

joshcooper replied:
Ah yeah good point. I updated defaults.rb to check for explicit checks instead of grepping with regexp

kris-bosland replied:
LGTM.

kris-bosland commented on lib/puppet/defaults.rb:
```diff
@@ -939,26 +945,37 @@ def self.default_checksum_types
+        :default  => lambda { default_digest_algorithm },
         :type     => :enum,
-        :values => ["md5", "sha256", "sha384", "sha512", "sha224"],
+        :values   => %w[md5 sha256 sha384 sha512 sha224],
```
How is :values used?  Should this also be set with a lambda that depends on Puppet::Util::Platform.fips_enabled? ?

joshcooper replied:
Yeah, I was thinking that too, but the issue is that puppet settings munges/validates the setting before hooks are called. So if I remove `md5` from the list when FIPS is enabled, and you try to use `puppet apply --digest_algorithm md5 -e ...`, then you'll get a mediocre message:

```
Error: Could not initialize global default settings: Invalid value 'md5' for parameter digest_algorithm. Allowed values are 'sha256', 'sha384', 'sha512', 'sha224'
```

I wanted to provide a more helpful message, which means I have to add `md5` to the list of values and do additional validation in the hook.

One alternative is to create a new settings type, though I'm not a fan of proliferation of new setting types.

Another option is to remove the hook logic and stick with the generic error message?

kris-bosland replied:
Can the user see the value of :values?  In what context?  I think what you have now is OK, as long as the user isn't confused by a list of options without clear documentation about the fips dependency to go along with it.

joshcooper replied:
With this PR the error message is:

```
Error: Could not initialize global default settings: MD5 digest is prohibited in FIPS mode. Valid values are sha256, sha384, sha512, sha224.
```

joshcooper replied:
Hmm, the new message doesn't specify which setting is causing the issue. Perhaps I should just drop that and use the generic message.

joshcooper commented:
I updated the error messages to be more consistent with the wording we use for invalid array values. Also updated the setting description text based on FIPS mode, eg so that `bx puppet doc --reference configuration` generates the correct values.

joshcooper commented:
There is an order dependent test failure in appveyor. The passing cell runs `23759 examples, 0 failures, 99 pending`. The failing cell `22769 examples, 0 failures, 94 pending`. I think the failing cell is exiting rspec, hence the missing 990 tests, similar to what we saw in https://github.com/puppetlabs/puppet/pull/6570#issuecomment-361364100.

joshcooper commented:
jenkins please test this on redhatfips7-64

kris-bosland commented:
👍

jhelwig commented on lib/puppet/defaults.rb:
```diff
@@ -939,26 +945,37 @@ def self.default_checksum_types
+        :values   => valid_digest_algorithms,
+        :desc     => "Which digest algorithm to use for file resources and the filebucket.
+                      Valid values are #{valid_digest_algorithms.join(', ')}. Default is
+                      #{valid_digest_algorithms.first}.",
```
Why is this `valid_digest_algorithms.first` instead of `default_digest_algorithm` (to match what's in the default lambda)?

jhelwig commented on lib/puppet/defaults.rb:
```diff
@@ -939,26 +945,37 @@ def self.default_checksum_types
-        invalid = values.reject {|alg| valid.include?(alg)}
+
+        if Puppet::Util::Platform.fips_enabled?
+          invalid = values.find { |v| v == 'md5' || v == 'md5lite' }
```
Isn't this effectively duplicating the invalid check below, since `Puppet.valid_checksum_types` won't contain `md5`, or `md5lite` when `Puppet::Util::Platform.fips_enabled?` is true? (Eg: `invalid = values - Puppet.valid_checksum_types`)
